### PR TITLE
Resolve #1167 -- Unhardcoded fountain

### LINKED ISSUE #1167: Unhardcode Fountations
Unhardcode the fountain in the SR file.

### PR DESCRIPTION
**Pull request by cabeca1143**
*Created a new AddMapObject function.
*Added an exception so when the room file doesn't contain the fountain files (as in Map1), it will grab these files directly.
*Unhardcoded the fountain the SR file.

Resolve #1167

[Removed: *Changed the enormous ammount of "else if"s in the buff section of AttackableUnit to a switch statement]

### CODE REVIEW

File: GameServerLib/GameObjects/AttackableUnits/AttackableUnit.cs [hunk 1]
```diff
@@ -588,159 +588,167 @@ public void AddBuff(IBuff b)
...
-                        return;
-                    }
-                    // If we haven't hit the max stack count (usually 254).
-                    BuffList.Add(b);
-
-                    // Increment the number of stacks on the parent buff, which is the buff instance which is used for packets.
-                    ParentBuffs[b.Name].IncrementStackCount();
+                        // If the buff is supposed to add a stack to any existing buffs of the same name and refresh their timer.
+                        // Essentially the method is: have one parent buff which has the stacks, and just refresh its time, this means no overlapping buff instances, but functionally it is the same.
+                        case BuffAddType.STACKS_AND_RENEWS:
```
Comment (danil179) [inline GameServerLib/GameObjects/AttackableUnits/AttackableUnit.cs]: In the original code it checked `ParentBuffs[b.Name].BuffAddType == BuffAddType.STACKS_AND_RENEWS` while here you check `b.BuffAddType`. I'm not sure what exactly it is changing, but the code is not doing exactly the same thing.

**Review by danil179 — CHANGES_REQUESTED** [review]
There is one problem in the buff switch statement. I also don't find the switch statement really more readable as there are too many conditions.

**Comment by cabeca1143** [comment]
> There is one problem in the buff switch statement. I also don't find the switch statement really more readable as there are too many conditions.

My intention wasn't really to make it more readable, just make so it doesn't check for every condition Everytime

File: GameServerLib/GameObjects/AttackableUnits/AttackableUnit.cs (hunk 1 above)
Comment (cabeca1143) [reply GameServerLib/GameObjects/AttackableUnits/AttackableUnit.cs]: An overlook by me.
I'll check that when i get back home

Comment (cabeca1143) [reply GameServerLib/GameObjects/AttackableUnits/AttackableUnit.cs]: From my testing, everything is working as it is supposed to.
ParentBuffs[b.Name] seems to be calling an Ibuff using a "b.name" string key, which will in turn just return "b" again.
Idk if I'm missing something, but it seems it was just redundant code.

**Comment by danil179** [comment]
> 
> 
> > There is one problem in the buff switch statement. I also don't find the switch statement really more readable as there are too many conditions.
> 
> My intention wasn't really to make it more readable, just make so it doesn't check for every condition Everytime

It doesn't check for every condition if you write `else if` everytime. Moreover, there are already numerous optimization when compiling that sort of try to predict what if to check first. From what I know there are no significant changes by changing ifs to switch. I suggest to just revert these changes to buffs.

**Comment by cabeca1143** [comment]
> > > There is one problem in the buff switch statement. I also don't find the switch statement really more readable as there are too many conditions.
> > 
> > 
> > My intention wasn't really to make it more readable, just make so it doesn't check for every condition Everytime
> 
> It doesn't check for every condition if you write `else if` everytime. Moreover, there are already numerous optimization when compiling that sort of try to predict what if to check first. From what I know there are no significant changes by changing ifs to switch. I suggest to just revert these changes to buffs.

Alright

**Review by danil179 — APPROVED** [review]
Approved. Seems to work with the spawn. There are existing problems with the fountains, so it is hard to verify all the details.

**Pull request merged**